cmd/admin-handlers.go

```diff
@@ -2790,41 +2813,81 @@ func (a adminAPIHandlers) InspectDataHandler(w http.ResponseWriter, r *http.Requ
 		return
 	}
 
+	var publicKey *rsa.PublicKey
+
+	publicKeyB64 := r.Form.Get("public-key")
+	if publicKeyB64 != "" {
+		publicKeyBytes, err := base64.StdEncoding.DecodeString(publicKeyB64)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		publicKey, err = bytesToPublicKey(publicKeyBytes)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+	}
+
 	var key [32]byte
 	// MUST use crypto/rand
 	n, err := crand.Read(key[:])
 	if err != nil || n != len(key) {
 		logger.LogIf(ctx, err)
-		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrInternalError), r.URL)
+		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 		return
 	}
-	stream, err := sio.AES_256_GCM.Stream(key[:])
+
+	// Write a version for making *incompatible* changes.
+	// The AdminClient will reject any version it does not know.
+	if publicKey == nil {
+		w.Write([]byte{1})
+		w.Write(key[:])
+	} else {
+		w.Write([]byte{2})
+	}
+
+	archiveW := zip.NewWriter(w)
+	defer archiveW.Close()
+
+	if publicKey != nil {
+		cipherKey, err := rsa.EncryptOAEP(sha512.New(), crand.Reader, publicKey, key[:], nil)
+		if err != nil {
+			// Too late to write an error over HTTP
+			logger.LogIf(ctx, err)
+			return
+		}
+		err = embedFileInZip(archiveW, "key.enc", cipherKey)
+		if err != nil {
+			logger.LogIf(ctx, err)
+			return
+		}
+	}
+
+	dataW, err := archiveW.CreateHeader(&zip.FileHeader{
+		Name: "data.zip",
+		// data.zip will be already compressed,
+		// no need to compress again this file.
+		Method: zip.Store,
+	})
```

Comment:
Instead of recursive ZIP (which is annoying to unpack) we can use a format similar to v1
```suggestion
	// Write a version for making *incompatible* changes.
	// The AdminClient will reject any version it does not know.
	if publicKey == nil {
		w.Write([]byte{1})
		w.Write(key[:])
	} else {
		// Version 2:
		// 0x2 (1 byte)
		// len(cipherKey) as little endian 32 bit value (4 bytes)
		// cipherKey
		// Encrypted ZIP data (until EOF)
		var header [5]byte
		header[0] = 2
		cipherKey, err := rsa.EncryptOAEP(sha512.New(), crand.Reader, publicKey, key[:], nil)
		if err != nil {
			writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrInternalError), r.URL)
			logger.LogIf(ctx, err)
			return
		}
		binary.LittleEndian.PutUint32(header[1:], uint32(len(cipherKey)))
		w.Write(header[:])
		w.Write(cipherKey)
	}
```